File: bindings/python/iota_sdk/types/irc_30.py
# Copyright 2023 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

import json
from dataclasses import dataclass, field, asdict
from typing import Optional
from dacite import from_dict
from iota_sdk.types.common import HexStr
from iota_sdk import utf8_to_hex, MetadataFeature


@dataclass
class Irc30Metadata:
    """The IRC30 native token metadata standard schema.
    Attributes:
        standard: The metadata standard (IRC30).
        name: The human-readable name of the native token.
        symbol: The symbol/ticker of the token.
        decimals: Number of decimals the token uses (divide the token amount by 10^decimals to get its user representation).
        description: The human-readable description of the token.
        url: URL pointing to more resources about the token.
        logoUrl: URL pointing to an image resource of the token logo.
        logo: The svg logo of the token encoded as a byte string.
    """

    standard: str = field(default="IRC30", init=False)
    name: str
    symbol: str
    decimals: int
    description: Optional[str] = None
    url: Optional[str] = None
    logoUrl: Optional[str] = None
    logo: Optional[HexStr] = None

    @staticmethod
    def from_dict(metadata_dict: dict):
        """
        Takes a dictionary as input and returns an instance of the `Irc30Metadata` class.
        """
        return from_dict(Irc30Metadata, metadata_dict)

    def as_hex(self):
        """Turns this schema into a hex encoded string
        """
        utf8_to_hex(json.dumps(asdict(self), separators=(",", ":")))

    def as_feature(self):
        """Turns this schema into a MetadataFeature type
        """
        MetadataFeature({'irc-30': self.as_hex()})

File: bindings/nodejs/lib/types/block/output/irc-30.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { utf8ToHex } from '../../../utils/utf8';
import { MetadataFeature } from './feature';

/**
 * The IRC30 native token metadata standard schema.
 */
class Irc30Metadata {
    /** The IRC standard */
    readonly standard: string = 'IRC30';
    /** The human-readable name of the native token. */
    name: string;
    /** The symbol/ticker of the token. */
    symbol: string;
    /** Number of decimals the token uses (divide the token amount by `10^decimals` to get its user representation). */
    decimals: number;
    /** The human-readable description of the token. */
    description?: string;
    /** URL pointing to more resources about the token. */
    url?: string;
    /** URL pointing to an image resource of the token logo. */
    logoUrl?: string;
    /** The svg logo of the token encoded as a byte string. */
    logo?: string;

    /**
     * @param name The human-readable name of the native token.
     * @param symbol The symbol/ticker of the token.
     * @param decimals Number of decimals the token uses.
     */
    constructor(name: string, symbol: string, decimals: number) {
        this.name = name;
        this.symbol = symbol;
        this.decimals = decimals;
    }

    withDescription(description: string): Irc30Metadata {
        this.description = description;
        return this;
    }

    withUrl(url: string): Irc30Metadata {
        this.url = url;
        return this;
    }

    withLogoUrl(logoUrl: string): Irc30Metadata {
        this.logoUrl = logoUrl;
        return this;
    }

    withLogo(logo: string): Irc30Metadata {
        this.logo = logo;
        return this;
    }

    asHex(): string {
        return utf8ToHex(JSON.stringify(this));
    }

    asFeature(): MetadataFeature {
        return new MetadataFeature(this.asHex());
    }
}

export { Irc30Metadata };
